wordpress-mobile/WordPressKit-iOS
# [PoC] WP-API and Cookie Authentication

### Description

We are currently lacking a way to perform requests to the [WordPress REST API](https://developer.wordpress.org/rest-api/) because we lack a good authentication method. The only method available today is based on [cookies and nonces](https://developer.wordpress.org/rest-api/using-the-rest-api/authentication/). There might be support for [Basic Authentication](https://core.trac.wordpress.org/ticket/42790) and [oAuth2](https://github.com/wp-api/authentication) eventually, but there's no timeline for that.

At the same time, we are coming across more and more features in gutenberg-mobile that require getting data from the API. In https://github.com/wordpress-mobile/WordPress-iOS/pull/12714 we managed to work around this issue because we could get enough information through an unauthenticated request. As part of https://github.com/wordpress-mobile/gutenberg-mobile/issues/1513 we will need to get the theme's color palette that will be exposed by https://core.trac.wordpress.org/ticket/48798, but only for authenticated users.

In this scenario, I think it makes sense to make an attempt at implementing a solution for the only available auth method, as fragile as that might be, so here it is.

This new API Client supports both Token and Cookie authentication, although only the latter has been tested so far (`GutenbergNetworkRequest` is already reusing the existing WordPress.com API Client for WordPress.com sites), but the basic design would allow for the other future authentication methods (detecting which one we should use might make things interesting, but that's not a problem for today). 

The `RequestAdapter` and `RequestRetrier` patterns that Alamofire offer seem pretty convenient for authentication, and if this moves forward, it would be interesting to refactor the existing `WordPressComRestApi` client, since there is a lot of overlap with the .org one, and a lot of the specifics could be configurable via adapters. 

### How it works

When an API request fails with a 401, the CookieNonce authenticator will POST to `wp-login.php` and ask to redirect to the new post page (`wp-admin/post-new.php`). The editor is the only page I've found where we can extract a usable nonce. Then the authenticator will look at the HTML source and extract the nonce from the Gutenberg initialization code. This also means the method won't work if they don't have Gutenberg as the default editor for new posts (pre-5.0 or classic editor plugin), but we don't have much of an alternative there.

Nonces have an expiration of 1 day, so instead of complicating things too much with storage and invalidation, I'm just keeping the nonce in memory as part of the Authenticator. This is tied to the lifetime of the API Client, which is tied to the Blog. I'm not checking for the presence of an existing nonce or cookie, so if a request fails because either expired, we would try to authenticate again and resolve the issue.

This also means that we still have to find a way to flag when the authentication fails. We don't want to be retrying the login on every API request if it's not going to work, and could lead to hosting providers blocking the user 😱 

### Testing Details

I've tested this from WPiOS https://github.com/wordpress-mobile/WordPress-iOS/pull/13020 and the steps described in https://github.com/wordpress-mobile/WordPress-iOS/pull/12714:

- Run metro server on develop
- Add a image block to gutenberg.
- Select an image from the WordPress media Library.
- Select settings and choose a different image setting.
- Change to HTML mode.
- Check that the image URL has changes according to the selected image size.
